File: java/src/yax/lex/Lexeme.java
/* Copyright 2009, UCAR/Unidata and OPeNDAP, Inc.
   See the COPYRIGHT file for more information. */

package yax.lex;

public class Lexeme
{
    public String name; // element or attribute name
    public int open; // Bison token for element open
    public int close; // Bison token for element close
    public int atoken; // Bison token if this is an attribute
    public String[] legalAttributes; /* not currently used */

    public Lexeme(String name, int atoken) {this(name,0,0,atoken,null);}
    public Lexeme(String name, int open, int close)
	{this(name,open,close,0,null);}
    public Lexeme(String name, int open, int close, String[] alist)
	{this(name,open,close,0,alist);}

    public Lexeme(String name, int open, int close, int atoken, String[] alist)
    {
	this.name = name;
	this.open = open;
	this.close = close;
	this.atoken = atoken;
	legalAttributes = alist;
    }

    public void attribute(String[] alist)
    {
	legalAttributes = alist;
	// check for duplicates
	for(int i=0;i<legalAttributes.length;i++) {
	    for(int j=i+1;j<legalAttributes.length;j++)
	        assert (!legalAttributes[i].equals(legalAttributes[j]));
	}
    }

    public String toString()
    {
         String text = String.format("%s open=%d close=%d a=%d",name,open,close,atoken);
         if(legalAttributes != null && legalAttributes.length > 0) {
             text += " attributes=";
             for(String s: legalAttributes) {
                 text += " " + s;
             }
         }
         return text;
    }
} // class Lexeme

File: python/yax/lexeme.py
# Copyright 2009, UCAR/Unidata
#   See the COPYRIGHT file for more information.

class Lexeme:

  def __init__(self, name=None, open=None, close=None, legal=None, attr=None) :
    self.name = name
    self.open = open
    self.close = close
    self.attr = attr
    self.setAttributes(legal)
  # end __init__

  def setAttributes(self,alist):
    self.legalAttributes = alist
    if(self.legalAttributes != None) :
      # check for duplicates
      for i in range(0,len(self.legalAttributes)) :
        for j in range(i+1,len(self.legalAttributes)) :
          assert (self.legalAttributes[i] != self.legalAttributes[j])
  #end attribute

  def __str__(self) :
    text = "{0} open={1} close={2} a={3}".format(self.name,str(self.open),str(self.close),str(self.attr))
    if(self.legalAttributes != None and len(self.legalAttributes) > 0) :
      text += " attributes="
      for s in self.legalAttributes :
        text += " " + s
    return text
  #end __str__

# end class Lexeme

